internal/api/v1beta1/app_types.go

```diff
@@ -665,3 +657,200 @@ func (t Target) IsDeployment() bool {
 func (t Target) IsService() bool {
 	return t.Kind == "Service" && t.APIVersion == "v1"
 }
+
+const (
+	CanaryNotActiveEvent     = "CanaryNotActive"
+	CanaryNotActiveEventDesc = "error - canary triggered, but not active"
+
+	CanaryNoDeployments     = "CanaryNoDeployments"
+	CanaryNoDeploymentsDesc = "error - canary needs more than 1 deployment to run"
+
+	CanaryNoScheduledSteps     = "CanaryNoScheduledSteps"
+	CanaryNoScheduledStepsDesc = "error - canary triggered, but no scheduled steps"
+
+	CanaryStarted      = "CanaryStarted"
+	CanaryStartedDesc  = "started"
+	CanaryFinished     = "CanaryFinished"
+	CanaryFinishedDesc = "finished"
+
+	CanaryNextStep       = "CanaryNextStep"
+	CanaryNextStepDesc   = "weight change"
+	CanaryStepTarget     = "CanaryStepTarget"
+	CanaryStepTargetDesc = "units change"
+)
+
+type CanaryEvent struct {
+	// AppName represents event for certain app
+	AppName string
+	// DeploymentVersion represents for which deployment event is accosted with
+	DeploymentVersion int
+
+	// Name represents canary event name. It is translated into Reason column of kubernetes event
+	// values: CanaryStarted, CanaryFinished
+	// errored values: CanaryNotActiveEvent, CanaryNoDeployments, CanaryNoScheduledSteps
+	Name string
+	// Description states what is the outcome of this event
+	Description string
+}
+
+func newCanaryEvent(app *App, event string, desc string) CanaryEvent {
+	var version DeploymentVersion
+	if len(app.Spec.Deployments) > 0 {
+		version = app.Spec.Deployments[len(app.Spec.Deployments)-1].Version
+	}
+
+	return CanaryEvent{
+		AppName:           app.Name,
+		DeploymentVersion: int(version),
+		Name:              event,
+		Description:       desc,
+	}
+}
+
+// Message is message for k8s event
+func (c CanaryEvent) Message() string {
+	return fmt.Sprintf("%s - Canary for app %s | version %d - %s", c.Name, c.AppName, c.DeploymentVersion, c.Description)
```

Comment:
I'd probably make `"%s - Canary for app %s | version %d - %s"`. Not a huge concern. Same with the NextStepEvents and ChangeEvents below.